File: c/sort/merge_sort.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

#define N 10
int a[N] = {1, 6, 3, 8, 15, 6, 7, 3, 12, 10};

void merge(int *a, int p, int q, int r) {

    int i = 0;
    int j = 0;
    int k = 0;

    int *tmp = malloc(sizeof(int) * (q-p+1));

    while ((i<(r-p+1)) || (j < (q-r))) {
        if (i == (r-p+1)) {
            tmp[k] = a[r+1+j];
            j++;
        } else if (j == (q-r)) {
            tmp[k] = a[p+i];
            i++;
        } else {
            if (a[p+i] < a[r+1+j]) {
                tmp[k] = a[p+i];
                i++;
            } else {
                tmp[k] = a[r+1+j];
                j++;
            }
        }
        k++;
    }

    for (int i = p; i < (q+1); i++)
        a[i] = tmp[i-p];
}

void merge_sort(int *a, int p, int q) {
    if (p >= q) {
        return;
    }

    int r = (int)((p+q)/2);
    merge_sort(a, p, r);
    merge_sort(a, r+1, q);
    merge(a, p, q, r);
}

int main() {

    merge_sort(a, 0, N-1);
    for (int i = 0; i < N; i++)
        printf("%d, ", a[i]);

    return 0;
}

File: python/sort/merge_sort.py

# a = [1, 4, 7, 3, 11, 9, 2, 6, 4, 7]
a = [2, 3, 3, 7, 7, 9, 10, 12, 20]
b = [1, 3, 4, 7, 10, 12, 12, 16, 17, 19]
t = a + b

def merge_sort(a, p, q):
    r = int((p + q) / 2)

    if p >= q:
        return

    merge_sort(a, p, r)
    merge_sort(a, r+1, q)
    merge(a, p, q, r)

# merge list m and n into a by ascending order
def merge(a, p, q, r):
    # i, j = p, r+1
    i = j = 0
    k = 0
    tmp = [None] * (q - p + 1)

    while i < (r-p+1) or j < (q-r):
        if (i == (r-p+1)):
            tmp[k] = a[r+1+j]
            j += 1
        elif (j == (q-r)):
            tmp[k] = a[p+i]
            i += 1
        else:
            if (a[p+i] < a[r+1+j]):
                tmp[k] = a[p+i]
                i += 1
            else:
                tmp[k] = a[r+1+j]
                j += 1

        k += 1

    print(tmp)

    for i in range(p, q+1):
        a[i] = tmp[i-p]

def main():
    # merge(t, 8, 9, 8)
    merge_sort(t, 0, len(t)-1)
    print(t)

if __name__ == '__main__':
    main()